akka-http2-support/src/main/scala/akka/http/impl/engine/http2/Http2ServerDemux.scala

```diff
@@ -204,7 +223,33 @@ class Http2ServerDemux extends GraphStage[BidiShape[Http2SubStream, FrameEvent,
           }
         }
       }
+
+      private def concatContinuationIntoHeaders(handler: SubStream, cont: ContinuationFrame) = {
+        val concatHeaderBlockFragment = handler.headers.headerBlockFragment ++ cont.payload
+        val moreCompleteHeadersFrame = HeadersFrame(handler.streamId, cont.endHeaders, cont.endHeaders, concatHeaderBlockFragment)
+        // update the SubStream we keep around
+        val moreCompleteHandler = handler.copy(headers = moreCompleteHeadersFrame)
+        incomingStreams += handler.streamId → moreCompleteHandler
+        moreCompleteHandler
+      }
+
+      def makeHttp2SubStream(handler: SubStream): Http2SubStream = {
+        val headers = handler.headers
+        val subStream = handler.outlet
+        if (headers.endStream && headers.endHeaders) {
+          Http2SubStream(headers, Source.empty)
+        } else {
+          // FIXME a bit naive but correct I think -- todo check the spec 
```

Comment:
Actually that potential sequence happens on the related stream not on the *actual* stream.
I actually made a mistake by calling it PUSH_STREAM, it's in fact `PUSH_PROMISE`.

Basically the PUSH_PROMISE can be sent after the data frame, basically I guess what you described is more what happens on the client them however the PUSH_PROMISE from the server is sent with the stream id of the connection anywhere between the headers and the end_stream and promises a stream id and then it will sequence it like you described.